core/src/apps/management/authenticate_device.py

```diff
@@ -0,0 +1,46 @@
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from trezor.messages import AuthenticateDevice, AuthenticityProof
+
+
+async def authenticate_device(msg: AuthenticateDevice) -> AuthenticityProof:
+    from trezor import utils
+    from trezor.crypto import optiga
+    from trezor.crypto.der import read_length
+    from trezor.crypto.hashlib import sha256
+    from trezor.messages import AuthenticityProof
+    from trezor.ui.layouts import confirm_action
+    from trezor.utils import BufferReader
+
+    from apps.common.writers import write_compact_size
+
+    await confirm_action(
+        "authenticate_device",
+        "Authenticate device",
+        description="Do you wish to verify the authenticity of your device?",
+    )
+
+    header = b"AuthenticateDevice:"
+    h = utils.HashWriter(sha256())
+    write_compact_size(h, len(header))
+    h.extend(header)
+    write_compact_size(h, len(msg.challenge))
+    h.extend(msg.challenge)
+    signature = optiga.sign(optiga.DEVICE_ECC_KEY_INDEX, h.get_digest())
+
+    certificates = []
+    r = BufferReader(optiga.get_certificate(optiga.DEVICE_CERT_INDEX))
+    while r.remaining_count() > 0:
+        cert_begin = r.offset
+        if r.get() != 0x30:
+            raise ValueError
```

Comment:
this converts to a generic FirmwareError on the wire. FirmwareError seems appropriate, but the text should be custom.
how about this?
```suggestion
            raise wire.FirmwareError("Device certificate is corrupted")
```

Reply:
https://github.com/trezor/trezor-firmware/pull/3255/commits/6928d6f24434a5e5721b608de059eedde7f356e0